### PR TITLE
Add rate limiter

### PR DESCRIPTION
**Pull request by algmyr**
Draft for the rate limiter. It needs to be integrated, and we should consider whether we want the lock as we currently do.

### CODE REVIEW

File: tle/util/codeforces_api.py [hunk 1]
```diff
@@ -183,30 +181,59 @@ def _bool_to_str(value):
...
+    async def wrapped(*args, **kwargs):
+        tries = 3
+        for i in range(tries):
+            now = time.time()
+
+            # Next valid slot is 1s after the `per_second`th last request
+            next_valid = max(now, 1 + last[0])
+            last.append(next_valid)
+            last.popleft()
+
```
Comment (cheran-senthil) [inline tle/util/codeforces_api.py]: A [queue](https://docs.python.org/3.7/library/queue.html) be a more accurate structure here.

Comment (algmyr) [reply tle/util/codeforces_api.py]: queue is for when synchronization is needed. IIRC for deque the relevant operations are atomic which makes the usage here fine, no need for synchronization.

File: tle/util/codeforces_api.py [hunk 2]
```diff
@@ -183,30 +181,59 @@ def _bool_to_str(value):
     raise TypeError(f'Expected bool, got {value} of type {type(value)}')
 
 
+def cf_ratelimit(f):
+    per_second = 5
+    last = deque([0]*per_second)
+    async def wrapped(*args, **kwargs):
```
Comment (meooow25) [inline tle/util/codeforces_api.py]: It's a good practice to use functools.wraps.

File: tle/util/codeforces_api.py [hunk 3]
```diff
@@ -183,30 +181,59 @@ def _bool_to_str(value):
...
+            if delay > 0:
+                await asyncio.sleep(delay)
+
+            try:
+                return await f(*args, **kwargs)
+            except (ClientError, CallLimitExceededError) as e:
+                if i < tries - 1:
+                    logger.info(f'Try {i+1}/{tries} at query failed. Retrying...')
+                else:
+                    logger.info(f'Try {i+1}/{tries} at query failed. Aborting.')
```
Comment (meooow25) [inline tle/util/codeforces_api.py]: We should log the error repr too. Also both messages are similar, we can print the common part once and "Retrying/Aborting" separately.

File: tle/util/codeforces_api.py [hunk 4]
```diff
@@ -183,30 +181,59 @@ def _bool_to_str(value):
...
+                    raise e
```
Comment (meooow25) [inline tle/util/codeforces_api.py]: Shouldn't there be a delay between tries? Especially when it's CallLimitExceeded, retrying immediately makes things worse.

Comment (algmyr) [reply tle/util/codeforces_api.py]: So, add some artificial delay if it's not the first instance of the loop?

File: tle/util/codeforces_api.py (hunk 2 above)
Comment (algmyr) [reply tle/util/codeforces_api.py]: I had no idea this was a thing, good to know.

**Comment by algmyr** [comment]
The discussions regarding this in a nutshell:
This is probably the simplest rate limiter that should do the job if CF respects their own call limits.
It will also behave nicely in an outage (failing quickly).
Adding stuff to it only means messing the idea of it up.
We can replace it with a better system later if needed, but this will do for now at least.

**Pull request merged**